File: src/game/TetrisGame.js
class TetrisGame {
  constructor(width = 10, height = 20) {
    this.width = width;
    this.height = height;
    this.board = this.createEmptyBoard();
    this.score = 0;
    this.level = 1;
    this.linesCleared = 0;
    this.gameOver = false;
    this.currentPiece = null;
    this.nextPiece = null;
    this.generateNewPiece();
  }

  createEmptyBoard() {
    return Array.from({ length: this.height }, () => Array(this.width).fill(0));
  }

  // Tetromino shapes (I, O, T, S, Z, J, L)
  static PIECES = {
    I: [[1, 1, 1, 1]],
    O: [
      [1, 1],
      [1, 1]
    ],
    T: [
      [0, 1, 0],
      [1, 1, 1]
    ],
    S: [
      [0, 1, 1],
      [1, 1, 0]
    ],
    Z: [
      [1, 1, 0],
      [0, 1, 1]
    ],
    J: [
      [1, 0, 0],
      [1, 1, 1]
    ],
    L: [
      [0, 0, 1],
      [1, 1, 1]
    ]
  };

  static PIECE_COLORS = {
    I: 1,
    O: 2,
    T: 3,
    S: 4,
    Z: 5,
    J: 6,
    L: 7
  };

  generateNewPiece() {
    const pieces = Object.keys(TetrisGame.PIECES);
    const randomPiece = pieces[Math.floor(Math.random() * pieces.length)];

    if (this.nextPiece) {
      this.currentPiece = this.nextPiece;
    } else {
      this.currentPiece = {
        shape: TetrisGame.PIECES[randomPiece],
        color: TetrisGame.PIECE_COLORS[randomPiece],
        x: Math.floor(this.width / 2) - 1,
        y: 0,
        type: randomPiece
      };
    }

    const nextRandomPiece = pieces[Math.floor(Math.random() * pieces.length)];
    this.nextPiece = {
      shape: TetrisGame.PIECES[nextRandomPiece],
      color: TetrisGame.PIECE_COLORS[nextRandomPiece],
      x: Math.floor(this.width / 2) - 1,
      y: 0,
      type: nextRandomPiece
    };

    if (this.checkCollision(this.currentPiece.shape, this.currentPiece.x, this.currentPiece.y)) {
      this.gameOver = true;
    }
  }

  checkCollision(shape, x, y) {
    for (let row = 0; row < shape.length; row++) {
      for (let col = 0; col < shape[row].length; col++) {
        if (shape[row][col]) {
          const newX = x + col;
          const newY = y + row;

          if (newX < 0 || newX >= this.width || newY >= this.height) {
            return true;
          }

          if (newY >= 0 && this.board[newY][newX]) {
            return true;
          }
        }
      }
    }
    return false;
  }

  rotate() {
    if (!this.currentPiece || this.gameOver) return false;

    const rotated = this.currentPiece.shape[0].map((_, i) =>
      this.currentPiece.shape.map(row => row[i]).reverse()
    );

    if (!this.checkCollision(rotated, this.currentPiece.x, this.currentPiece.y)) {
      this.currentPiece.shape = rotated;
      return true;
    }
    return false;
  }

  moveLeft() {
    if (!this.currentPiece || this.gameOver) return false;

    if (!this.checkCollision(this.currentPiece.shape, this.currentPiece.x - 1, this.currentPiece.y)) {
      this.currentPiece.x--;
      return true;
    }
    return false;
  }

  moveRight() {
    if (!this.currentPiece || this.gameOver) return false;

    if (!this.checkCollision(this.currentPiece.shape, this.currentPiece.x + 1, this.currentPiece.y)) {
      this.currentPiece.x++;
      return true;
    }
    return false;
  }

  moveDown() {
    if (!this.currentPiece || this.gameOver) return false;

    if (!this.checkCollision(this.currentPiece.shape, this.currentPiece.x, this.currentPiece.y + 1)) {
      this.currentPiece.y++;
      return true;
    } else {
      this.lockPiece();
      return false;
    }
  }

  hardDrop() {
    if (!this.currentPiece || this.gameOver) return;

    while (!this.checkCollision(this.currentPiece.shape, this.currentPiece.x, this.currentPiece.y + 1)) {
      this.currentPiece.y++;
    }
    this.lockPiece();
  }

  lockPiece() {
    const { shape, x, y, color } = this.currentPiece;

    for (let row = 0; row < shape.length; row++) {
      for (let col = 0; col < shape[row].length; col++) {
        if (shape[row][col]) {
          const boardY = y + row;
          const boardX = x + col;
          if (boardY >= 0) {
            this.board[boardY][boardX] = color;
          }
        }
      }
    }

    this.clearLines();
    this.generateNewPiece();
  }

  clearLines() {
    let linesCleared = 0;

    for (let row = this.height - 1; row >= 0; row--) {
      if (this.board[row].every(cell => cell !== 0)) {
        this.board.splice(row, 1);
        this.board.unshift(Array(this.width).fill(0));
        linesCleared++;
        row++; // Check the same row again
      }
    }

    if (linesCleared > 0) {
      this.linesCleared += linesCleared;
      this.score += this.calculateScore(linesCleared);
      this.level = Math.floor(this.linesCleared / 10) + 1;
    }
  }

  calculateScore(lines) {
    const baseScores = [0, 100, 300, 500, 800];
    return baseScores[lines] * this.level;
  }

  getState() {
    return {
      board: this.board.map(row => [...row]),
      currentPiece: this.currentPiece ? {
        shape: this.currentPiece.shape,
        x: this.currentPiece.x,
        y: this.currentPiece.y,
        color: this.currentPiece.color,
        type: this.currentPiece.type
      } : null,
      nextPiece: this.nextPiece ? {
        shape: this.nextPiece.shape,
        color: this.nextPiece.color,
        type: this.nextPiece.type
      } : null,
      score: this.score,
      level: this.level,
      linesCleared: this.linesCleared,
      gameOver: this.gameOver
    };
  }

  reset() {
    this.board = this.createEmptyBoard();
    this.score = 0;
    this.level = 1;
    this.linesCleared = 0;
    this.gameOver = false;
    this.currentPiece = null;
    this.nextPiece = null;
    this.generateNewPiece();
  }
}

module.exports = TetrisGame;

File: android/app/src/main/java/com/tetris/game/TetrisGame.kt
package com.tetris.game

data class GameState(
    val board: Array<IntArray>,
    val currentPiece: Piece?,
    val nextPiece: Piece?,
    val score: Int,
    val level: Int,
    val linesCleared: Int,
    val gameOver: Boolean
)

data class Piece(
    val shape: Array<IntArray>,
    val x: Int,
    val y: Int,
    val color: Int,
    val type: String
)

class TetrisGame(
    private val width: Int = 10,
    private val height: Int = 20
) {
    var board: Array<IntArray> = createEmptyBoard()
        private set

    var score: Int = 0
        private set

    var level: Int = 1
        private set

    var linesCleared: Int = 0
        private set

    var gameOver: Boolean = false
        private set

    var currentPiece: Piece? = null
        private set

    var nextPiece: Piece? = null
        private set

    companion object {
        private val PIECES = mapOf(
            "I" to arrayOf(intArrayOf(1, 1, 1, 1)),
            "O" to arrayOf(
                intArrayOf(1, 1),
                intArrayOf(1, 1)
            ),
            "T" to arrayOf(
                intArrayOf(0, 1, 0),
                intArrayOf(1, 1, 1)
            ),
            "S" to arrayOf(
                intArrayOf(0, 1, 1),
                intArrayOf(1, 1, 0)
            ),
            "Z" to arrayOf(
                intArrayOf(1, 1, 0),
                intArrayOf(0, 1, 1)
            ),
            "J" to arrayOf(
                intArrayOf(1, 0, 0),
                intArrayOf(1, 1, 1)
            ),
            "L" to arrayOf(
                intArrayOf(0, 0, 1),
                intArrayOf(1, 1, 1)
            )
        )

        private val PIECE_COLORS = mapOf(
            "I" to 1, "O" to 2, "T" to 3,
            "S" to 4, "Z" to 5, "J" to 6, "L" to 7
        )
    }

    init {
        generateNewPiece()
    }

    private fun createEmptyBoard(): Array<IntArray> {
        return Array(height) { IntArray(width) { 0 } }
    }

    fun generateNewPiece() {
        val pieces = PIECES.keys.toList()
        val randomPiece = pieces.random()

        if (nextPiece != null) {
            currentPiece = nextPiece
        } else {
            currentPiece = Piece(
                shape = PIECES[randomPiece]!!.map { it.clone() }.toTypedArray(),
                x = width / 2 - 1,
                y = 0,
                color = PIECE_COLORS[randomPiece]!!,
                type = randomPiece
            )
        }

        val nextRandomPiece = pieces.random()
        nextPiece = Piece(
            shape = PIECES[nextRandomPiece]!!.map { it.clone() }.toTypedArray(),
            x = width / 2 - 1,
            y = 0,
            color = PIECE_COLORS[nextRandomPiece]!!,
            type = nextRandomPiece
        )

        currentPiece?.let {
            if (checkCollision(it.shape, it.x, it.y)) {
                gameOver = true
            }
        }
    }

    private fun checkCollision(shape: Array<IntArray>, x: Int, y: Int): Boolean {
        for (row in shape.indices) {
            for (col in shape[row].indices) {
                if (shape[row][col] != 0) {
                    val newX = x + col
                    val newY = y + row

                    if (newX < 0 || newX >= width || newY >= height) {
                        return true
                    }

                    if (newY >= 0 && board[newY][newX] != 0) {
                        return true
                    }
                }
            }
        }
        return false
    }

    fun rotate(): Boolean {
        val piece = currentPiece ?: return false
        if (gameOver) return false

        val rotated = Array(piece.shape[0].size) { i ->
            IntArray(piece.shape.size) { j ->
                piece.shape[piece.shape.size - 1 - j][i]
            }
        }

        if (!checkCollision(rotated, piece.x, piece.y)) {
            currentPiece = piece.copy(shape = rotated)
            return true
        }
        return false
    }

    fun moveLeft(): Boolean {
        val piece = currentPiece ?: return false
        if (gameOver) return false

        if (!checkCollision(piece.shape, piece.x - 1, piece.y)) {
            currentPiece = piece.copy(x = piece.x - 1)
            return true
        }
        return false
    }

    fun moveRight(): Boolean {
        val piece = currentPiece ?: return false
        if (gameOver) return false

        if (!checkCollision(piece.shape, piece.x + 1, piece.y)) {
            currentPiece = piece.copy(x = piece.x + 1)
            return true
        }
        return false
    }

    fun moveDown(): Boolean {
        val piece = currentPiece ?: return false
        if (gameOver) return false

        if (!checkCollision(piece.shape, piece.x, piece.y + 1)) {
            currentPiece = piece.copy(y = piece.y + 1)
            return true
        } else {
            lockPiece()
            return false
        }
    }

    fun hardDrop() {
        val piece = currentPiece ?: return
        if (gameOver) return

        var newY = piece.y
        while (!checkCollision(piece.shape, piece.x, newY + 1)) {
            newY++
        }
        currentPiece = piece.copy(y = newY)
        lockPiece()
    }

    private fun lockPiece() {
        val piece = currentPiece ?: return

        for (row in piece.shape.indices) {
            for (col in piece.shape[row].indices) {
                if (piece.shape[row][col] != 0) {
                    val boardY = piece.y + row
                    val boardX = piece.x + col
                    if (boardY >= 0) {
                        board[boardY][boardX] = piece.color
                    }
                }
            }
        }

        clearLines()
        generateNewPiece()
    }

    private fun clearLines() {
        var cleared = 0

        var row = height - 1
        while (row >= 0) {
            if (board[row].all { it != 0 }) {
                // Remove line and add empty line at top
                val newBoard = Array(height) { IntArray(width) }
                for (i in 0 until row) {
                    newBoard[i + 1] = board[i]
                }
                board = newBoard
                cleared++
            } else {
                row--
            }
        }

        if (cleared > 0) {
            linesCleared += cleared
            score += calculateScore(cleared)
            level = linesCleared / 10 + 1
        }
    }

    private fun calculateScore(lines: Int): Int {
        val baseScores = listOf(0, 100, 300, 500, 800)
        return baseScores[lines] * level
    }

    fun getState(): GameState {
        return GameState(
            board = board.map { it.clone() }.toTypedArray(),
            currentPiece = currentPiece,
            nextPiece = nextPiece,
            score = score,
            level = level,
            linesCleared = linesCleared,
            gameOver = gameOver
        )
    }

    fun reset() {
        board = createEmptyBoard()
        score = 0
        level = 1
        linesCleared = 0
        gameOver = false
        currentPiece = null
        nextPiece = null
        generateNewPiece()
    }
}
